domain/schema/model.go

```diff
@@ -325,3 +341,99 @@ CREATE UNIQUE INDEX idx_unit_net_node
+    -- TODO (manadart 2023-11-29): Storage pools in Mongo were stored as settings.
+    -- The types are sourced from the provider.
+    -- We can:
+    --   1. Leave it as text on the basis that we don't know up front what types the provider offers.
+    --   2. Make a lookup with all the known types, which would need maintenance over the long run.
+    --   3. Populate a lookup at model creation with the provider's known types.
+    -- Option 1 preserves the status quo.
+    type TEXT NOT NULL
+);
+
+CREATE TABLE storage_pool_attributes (
+    storage_pool_uuid TEXT NOT NULL,
+    key               TEXT,
+    value             TEXT,
+    CONSTRAINT       fk_storage_pool_attribute_pool
+        FOREIGN KEY  (storage_pool_uuid)
+        REFERENCES   storage_pool(uuid)
+);
+
+CREATE UNIQUE INDEX idx_storage_pool_attribute
+ON storage_pool_attributes (storage_pool_uuid, key);
+
+CREATE TABLE storage_type (
+    id   INT PRIMARY KEY,
+    type TEXT
+);
+
+CREATE UNIQUE INDEX idx_storage_type_type
+ON storage_type (type);
+
+INSERT INTO storage_type VALUES
+    (0, 'block'), 
+    (1, 'filesystem');
+
+CREATE TABLE storage_instance (
+    uuid            TEXT PRIMARY KEY,
+    storage_type_id INT NOT NULL,
+    name            TEXT NOT NULL,
+    life_id         INT NOT NULL,
+    CONSTRAINT       fk_storage_instance_type
+        FOREIGN KEY  (storage_type_id)
+        REFERENCES   storage_type(id),
+    CONSTRAINT       fk_storage_instance_life
+        FOREIGN KEY  (life_id)
+        REFERENCES   life(id)
+);
+
+CREATE TABLE storage_instance_pool (
+    storage_instance_uuid TEXT PRIMARY KEY,
+    storage_pool_uuid     TEXT NOT NULL,
+    CONSTRAINT       fk_storage_instance_pool_instance
+        FOREIGN KEY  (storage_instance_uuid)
+        REFERENCES   storage_instance(uuid),
+    CONSTRAINT       fk_storage_instance_pool_pool
+        FOREIGN KEY  (storage_pool_uuid)
+        REFERENCES   storage_pool(uuid)
+);
+
+-- storage_unit_owner is used to indicate when
+-- a unit is the owner of a storage instance.
+-- This is different to a storage attachment.
+CREATE TABLE storage_unit_owner (
+    storage_instance_uuid TEXT PRIMARY KEY,
+    unit_uuid             TEXT NOT NULL,
+    CONSTRAINT       fk_storage_owner_storage
+        FOREIGN KEY  (storage_instance_uuid)
+        REFERENCES   storage_instance(uuid),
+    CONSTRAINT       fk_storage_owner_unit
+        FOREIGN KEY  (unit_uuid)
+        REFERENCES   unit(uuid)
+);
+
+-- Should owner just be a boolean on this table,
+-- or can a unit be an owner without being attached?
+CREATE TABLE storage_attachment (
+    storage_instance_uuid TEXT PRIMARY KEY,
+    unit_uuid             TEXT NOT NULL,
+    CONSTRAINT       fk_storage_owner_storage
+        FOREIGN KEY  (storage_instance_uuid)
+        REFERENCES   storage_instance(uuid),
+    CONSTRAINT       fk_storage_owner_unit
+        FOREIGN KEY  (unit_uuid)
+        REFERENCES   unit(uuid)
+);
+
+-- Does an instance always have constraints?
```

Comment:
Each cloud provider has a (provider specific) default storage provider type eg loop, ebs, rootfs etc. So this constraint attribute always has a value. And we also use a common default for size and count. So there is an implicit constraint always.